src/simulators/density_matrix/densitymatrix_state.hpp

```diff
@@ -518,6 +524,113 @@ void State<densmat_t>::snapshot_pauli_expval(const Operations::Op &op,
                             expval, variance);
 }
 
+
+template <class densmat_t>
+void State<densmat_t>::snapshot_density_matrix(const Operations::Op &op,
+                                               ExperimentData &data) {
+  cmatrix_t reduced_state;
+
+  // Check if tracing over all qubits
+  if (op.qubits.empty()) {
+    reduced_state = cmatrix_t(1, 1);
+    reduced_state[0] = BaseState::qreg_.trace();
+  } else {
+
+    auto qubits_sorted = op.qubits;
+    std::sort(qubits_sorted.begin(), qubits_sorted.end());
+
+    if ((op.qubits.size() == BaseState::qreg_.num_qubits()) && (op.qubits == qubits_sorted)) {
+      reduced_state = std::move(BaseState::qreg_.matrix());
```

Comment:
`qreg_.matrix` returns a copy, not a reference so there this isn't an issue with underfined state being left behind. Maybe move is unnecessary here like you mention below because of RVO